File: src/main/java/me/hhitt/disasters/disaster/impl/ExplosiveSheep.java
package me.hhitt.disasters.disaster.impl;

import me.hhitt.disasters.arena.Arena;
import me.hhitt.disasters.disaster.Disaster;
import me.hhitt.disasters.model.entity.DisasterSheep;
import me.hhitt.disasters.util.Notify;
import me.hhitt.disasters.util.SpawnLocationFinder;
import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.entity.ArmorStand;
import org.bukkit.entity.Player;
import org.bukkit.entity.Sheep;
import org.bukkit.inventory.ItemStack;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.CopyOnWriteArrayList;

public class ExplosiveSheep implements Disaster {

    private final Map<Arena, CopyOnWriteArrayList<DisasterSheep>> arenaSheep = new HashMap<>();

    @Override
    public void start(Arena arena) {
        arenaSheep.put(arena, new CopyOnWriteArrayList<DisasterSheep>());
        Notify.disaster(arena, "explosive-sheep");
    }

    @Override
    public void pulse(int time) {
        tick();

        if (time % 5 != 0) return;
        for (Arena arena : new ArrayList<>(arenaSheep.keySet())) {
            for (Player player : arena.getAlive()) {
                spawnSheep(arena, player, 10, 1);
            }
        }
    }

    @Override
    public void stop(Arena arena) {
        CopyOnWriteArrayList<DisasterSheep> removed = arenaSheep.remove(arena);
        if (removed != null) {
            for (DisasterSheep sheep : removed) {
                sheep.remove();
            }
        }
    }

    private void tick() {
        for (CopyOnWriteArrayList<DisasterSheep> sheeps : arenaSheep.values()) {
            for (DisasterSheep sheep : new ArrayList<DisasterSheep>(sheeps)) {
                if (sheep.isAlive()) {
                    sheep.call();
                } else {
                    sheeps.remove(sheep);
                }
            }
        }
    }

    private void spawnSheep(Arena arena, Player player, int radius, int amount) {
        for (int i = 0; i < amount; i++) {
            Location spawnLocation = SpawnLocationFinder.findNearPlayer(arena, player.getLocation(), 3, radius, 5);
            if (spawnLocation.getWorld() == null) continue;
            Sheep bukkitSheep = spawnLocation.getWorld().spawn(spawnLocation, Sheep.class);
            DisasterSheep sheep = new DisasterSheep(arena, bukkitSheep);
            addTntDisplay(sheep);
            CopyOnWriteArrayList<DisasterSheep> list = arenaSheep.get(arena);
            if (list != null) {
                list.add(sheep);
            }
        }
    }

    private void addTntDisplay(DisasterSheep sheep) {
        Sheep entity = sheep.getSheep();
        ArmorStand stand = entity.getWorld().spawn(entity.getLocation(), ArmorStand.class);
        stand.setVisible(false);
        stand.setSmall(true);
        stand.setMarker(true);
        stand.setInvulnerable(true);
        stand.setGravity(false);
        stand.getEquipment().setHelmet(new ItemStack(Material.TNT));
        entity.addPassenger(stand);
    }
}

File: src/main/kotlin/me/hhitt/disasters/disaster/impl/ExplosiveSheep.kt
package me.hhitt.disasters.disaster.impl

import me.hhitt.disasters.arena.Arena
import me.hhitt.disasters.disaster.Disaster
import me.hhitt.disasters.model.entity.DisasterSheep
import me.hhitt.disasters.util.Notify
import me.hhitt.disasters.util.SpawnLocationFinder
import net.minecraft.world.entity.Entity
import org.bukkit.Location
import org.bukkit.craftbukkit.CraftWorld
import org.bukkit.entity.Player
import java.util.concurrent.CopyOnWriteArrayList

class ExplosiveSheep: Disaster {

    val arenaSheep = mutableMapOf<Arena, CopyOnWriteArrayList<DisasterSheep>>()

    override fun start(arena: Arena) {
        arenaSheep[arena] = CopyOnWriteArrayList()
        Notify.disaster(arena, "explosive-sheep")
    }

    override fun pulse(time: Int) {
        tick()

        if(time % 5 != 0) return
        arenaSheep.keys.toList().forEach { arena ->
            arena.alive.forEach { player ->
                spawnSheep(arena, player, 10, 1)
            }
        }
    }

    override fun stop(arena: Arena) {
        arenaSheep.remove(arena)?.forEach { it.remove(Entity.RemovalReason.KILLED) }
    }

    private fun tick() {
        arenaSheep.values.forEach { sheeps ->
            sheeps.removeIf { sheep ->
                if (sheep.isAlive) {
                    sheep.call()
                    false
                } else {
                    true
                }
            }
        }
    }

    private fun spawnSheep(arena: Arena, player: Player, radius: Int, amount: Int) {
        repeat(amount) {
            val spawnLocation: Location = SpawnLocationFinder.findNearPlayer(arena, player.location, 3, radius, 5)
            val handle = (spawnLocation.world as CraftWorld).handle
            val sheep = DisasterSheep(net.minecraft.world.entity.EntityType.SHEEP, handle.level, spawnLocation)
            handle.addFreshEntity(sheep)
            arenaSheep[arena]?.add(sheep)
        }
    }
}